include/cetl/any.hpp

```diff
@@ -0,0 +1,766 @@
+        {
+            return;
+        }
+
+        if (has_value())
+        {
+            if (rhs.has_value())
+            {
+                any tmp{std::move(rhs)};
+                static_cast<base&>(rhs)   = std::move(*this);
+                static_cast<base&>(*this) = std::move(tmp);
+            }
+            else
+            {
+                static_cast<base&>(rhs) = std::move(*this);
+            }
+        }
+        else if (rhs.has_value())
+        {
+            static_cast<base&>(*this) = std::move(rhs);
+        }
+    }
+
+    CETL_NODISCARD bool has_value() const noexcept
+    {
+        return base::has_value();
+    }
+
+private:
+    template <typename ValueType, typename Any>
+    friend std::add_pointer_t<ValueType> any_cast(Any* operand) noexcept;
+
+    template <typename ValueType, typename Any>
+    friend std::add_pointer_t<std::add_const_t<ValueType>> any_cast(const Any* operand) noexcept;
+
+    template <typename Tp, typename... Args>
+    Tp& create(Args&&... args)
+    {
+        base::template make_handlers<Tp>();
+        return *new (base::get_raw_storage()) Tp(std::forward<Args>(args)...);
+    }
+
+};  // class any
+
+/// \brief Typealias for `any` with the given `ValueType` with the default
+/// footprint, copyability, movability, and alignment of the `ValueType`.
+///
+/// In use by `cetl::make_any` overloads to make them close to `std::make_any`.
+///
+template <typename ValueType>
+using any_like = any<sizeof(ValueType),
+                     std::is_copy_constructible<ValueType>::value,
+                     std::is_move_constructible<ValueType>::value,
+                     alignof(ValueType)>;
+
+/// \brief Constructs an any object containing an object of type T, passing the provided arguments to T's constructor.
+///
+/// Equivalent to `cetl::any(cetl::in_place_type<ValueType>, std::forward<Args>(args)...)`.
+///
+template <typename ValueType, typename Any = any_like<ValueType>, typename... Args>
+CETL_NODISCARD Any make_any(Args&&... args)
+{
+    return Any(in_place_type<ValueType>, std::forward<Args>(args)...);
+}
+
+/// \brief Constructs an any object containing an object of type T, passing the provided arguments to T's constructor.
+///
+/// Equivalent to `cetl::any(cetl::in_place_type<ValueType>, list, std::forward<Args>(args)...)`.
+///
+template <typename ValueType, typename Any = any_like<ValueType>, typename Up, typename... Args>
+CETL_NODISCARD Any make_any(std::initializer_list<Up> list, Args&&... args)
+{
+    return Any(in_place_type<ValueType>, list, std::forward<Args>(args)...);
+}
+
+/// \brief Performs type-safe access to the contained object.
+///
+/// \param operand Target any object.
+/// \return Returns `std::static_cast<ValueType>(*cetl::any_cast<const U>(&operand))`,
+///     where let `U` be `std::remove_cv_t<std::remove_reference_t<ValueType>>`.
+///
+template <typename ValueType, typename Any>
+CETL_NODISCARD ValueType any_cast(const Any& operand)
```

Comment:
We should refuse to generate any_cast for non-pointer types when exceptions are disabled. This is the approach we took with variable_length_array.hpp

Reply:
I'm expecting that in user code it will result in this:

```c++
T& val = any_cast<T>(my_any);
```

Being replaced with this:

```c++
T* const val = any_cast<T*>(&my_any);
assert(val != nullptr);
```

Where there is a robust assumption that the `any` contains `T`. One example would be the `DynamicBuffer` in LibCyphal.

Is it done to uphold compliance with a coding standard or there is other rationale?

Reply:
Of my two rules for API design: 
1. Make the API easy to use correctly
2. Make the API hard to use incorrectly

rule 2 applies here.

Reply:
Okay. But for the record we have throwing overloads in `cetl::variant`, like `cetl::get`.

Reply:
I renamed `any_cast` to two functions:
- → `get` for value-returning; throws `cetl::bad_unbounding_variant_access` in case of failure
- → `get_if` for pointer-returning; returns `nullptr` in case of failure
So now, this api is close to usual (type-bound) `cetl::variant`.